flight/Modules/DacBeepCodes/dacbeepcodes.c

```diff
@@ -0,0 +1,129 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "openpilot.h"
+#include "pios_thread.h"
+#include "pios_dacbeep_priv.h"
+
+#include "flightbatterysettings.h"
+#include "flightbatterystate.h"
+#include "manualcontrolcommand.h"
+
+#define STACK_SIZE_BYTES 648
+#define TASK_PRIORITY PIOS_THREAD_PRIO_LOW
+
+// Private variables
+static struct pios_thread *dacBeepCodeTaskHandle;
+static bool module_enabled;
+
+// Private functions
+static void dacBeepCodeTask(void *parameters);
+
+// External variables
+extern uintptr_t dacbeep_handle;
+
+ /**
+ * Initialise the module, called on startup
+ * \returns 0 on success or -1 if initialisation failed
+ */
+int32_t DacBeepCodesStart()
+{
+	// Start main task if it is enabled
+	if (module_enabled) {
+		dacBeepCodeTaskHandle = PIOS_Thread_Create(dacBeepCodeTask, "DacBeepCodes", STACK_SIZE_BYTES, NULL, TASK_PRIORITY);
+		TaskMonitorAdd(TASKINFO_RUNNING_ALTITUDEHOLD, dacBeepCodeTaskHandle);
+		return 0;
+	}
+	return -1;
+
+}
+
+/**
+ * Initialise the module, called on startup
+ * \returns 0 on success or -1 if initialisation failed
+ */
+int32_t DacBeepCodesInitialize()
+{
+	if (dacbeep_handle)
+		module_enabled = true;
+
+	if(module_enabled) {
+
+		return 0;
+	}
+
+	return -1;
+}
+
+MODULE_INITCALL(DacBeepCodesInitialize, DacBeepCodesStart);
+
+/**
+ * Module thread, should not return.
+ */
+static void dacBeepCodeTask(void *parameters)
+{
+	while(1) {
+
+		int16_t rssi = 0;
+		ManualControlCommandRssiGet(&rssi);
+		if (rssi < 0) rssi = 0;
+		if (rssi > 100) rssi = 100;
+
+		// Determine frequency from battery voltage
+		float volt = 0.0f;
+		if (FlightBatteryStateHandle())
+			FlightBatteryStateVoltageGet(&volt);
+
+		// We want to map RSSI inversely into pulse duration. At 100% shoot for
+		// a period of 800 ms. At 50% get to 150ms. Sharp inflection at 50%.
+		// Stops at 100 ms to set max rep rate at 10 Hz
+		// Using piecewise linear to avoid exponential calculation
+		int32_t period_ms = 100 + ((rssi < 50) ? rssi : (13 * (rssi - 50) + 50));
+
+		// Scale audio pitches based on voltage thresholds in ALARM and WARNING
+		// this makes the audio start to change once you hit this range and have
+		// a fair bit of dynamic range
+		float volt_threshold[FLIGHTBATTERYSETTINGS_VOLTAGETHRESHOLDS_NUMELEM];
+		FlightBatterySettingsVoltageThresholdsGet(volt_threshold);
+		const float MIN_VOLT = volt_threshold[FLIGHTBATTERYSETTINGS_VOLTAGETHRESHOLDS_ALARM];
+		const float MAX_VOLT = volt_threshold[FLIGHTBATTERYSETTINGS_VOLTAGETHRESHOLDS_WARNING];
+		const uint16_t MIN_FREQ_HZ = 1500;
+		const uint16_t RANGE_FREQ_HZ = 4000;
```

Comment:
I'll merge as-is, but do you think it makes more sense to set a max and min frequency, instead of a min frequency and then range? Or, maybe even makes sense to define this in octaves, since most people won't realize that 1500Hz-5500Hz is much more compact than 150Hz-4150Hz?